File: leetcode-practice-typescript/src/two-pointers/392-is-subsequence.ts
function isSubsequence(s: string, t: string): boolean {
    if (s.length > t.length) {
        return false;
    }
    let p1 = 0;
    let p2 = 0;
    while (p1 < s.length && p2 < t.length) {
        if (s[p1] !== t[p2]) {
            p2++;
            continue;
        }
        p1++;
        p2++;
    }
    return p1 === s.length;
}

console.log(isSubsequence('abc', 'ahbgdc'));
console.log(isSubsequence('axc', 'ahbgdc'));

File: leetcode-practice-python/src/two_pointers/392_is_subsequence.py
class Solution:
    def is_subsequence(self, s: str, t: str) -> bool:
        if not s:
            return True
        p1, p2 = 0, 0
        while p1 < len(s) and p2 < len(t):
            while p2 < len(t) and s[p1] != t[p2]:
                p2 += 1
            if p2 == len(t):
                break
            p1 += 1
            p2 += 1
        return p1 == len(s)


if __name__ == "__main__":
    s = Solution()
    print(s.is_subsequence("abc", "ahbgdc"))
    print(s.is_subsequence("axc", "ahbgdc"))
